onnxruntime/core/providers/cpu/sequence/sequence_ops.cc

```diff
@@ -0,0 +1,147 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/providers/cpu/sequence/sequence_ops.h"
+#include "core/framework/tensorprotoutils.h"
+
+using namespace onnxruntime::common;
+
+namespace onnxruntime {
+
+// SequenceLength
+ONNX_CPU_OPERATOR_KERNEL(
+    SequenceLength,
+    11,
+    KernelDefBuilder()
+        .TypeConstraint("S", DataTypeImpl::AllSequenceTensorTypes())
+        .TypeConstraint("I", DataTypeImpl::GetTensorType<int64_t>()),
+    SequenceLength);
+
+Status SequenceLength::Compute(OpKernelContext* context) const {
+  const auto* X = context->Input<VectorTensor>(0);
+  ORT_ENFORCE(X != nullptr, "Got nullptr for sequence input.");
+
+  auto* Y = context->Output(0, {});
+  auto* Y_data = Y->template MutableData<int64_t>();
+  *Y_data = static_cast<int64_t>(X->size());
+
+  return Status::OK();
+}
+
+// SequenceAt
+ONNX_CPU_OPERATOR_KERNEL(
+    SequenceAt,
+    11,
+    KernelDefBuilder()
+        .TypeConstraint("S", DataTypeImpl::AllSequenceTensorTypes())
+        .TypeConstraint("T", DataTypeImpl::AllTensorTypes())
+        .TypeConstraint("I", std::vector<MLDataType>{
+                                 DataTypeImpl::GetTensorType<int32_t>(),
+                                 DataTypeImpl::GetTensorType<int64_t>()}),
+    SequenceAt);
+
+static void GetSeqIdx(const Tensor& idx_tensor, int64_t& seq_idx) {
+  auto idx_tensor_dtype = utils::GetTensorProtoType(idx_tensor);
+  switch (idx_tensor_dtype) {
+    case ONNX_NAMESPACE::TensorProto_DataType_INT32: {
+      const auto* idx_data = idx_tensor.Data<int32_t>();
+      seq_idx = static_cast<int64_t>(*idx_data);
+      break;
+    }
+    case ONNX_NAMESPACE::TensorProto_DataType_INT64: {
+      const auto* idx_data = idx_tensor.Data<int64_t>();
+      seq_idx = *idx_data;
+      break;
+    }
+    default:
+      ORT_THROW("Unsupported data type: ", idx_tensor_dtype);
+  }
+}
+
+bool ValidateSeqIdx(int64_t input_seq_idx, int64_t seq_size) {
+  bool retval = false;
+  if (input_seq_idx < 0) {
+    retval = input_seq_idx <= -1 && input_seq_idx >= -seq_size;
+  } else {
+    retval = input_seq_idx < seq_size;
+  }
+  return retval;
+}
+
+template <typename T>
+static void CopyTensor(const Tensor& indexed_tensor, Tensor& output_tensor) {
+  const auto* input_data = indexed_tensor.template Data<T>();
```

Comment:
Can you use CopyCpuTensor from providers\cpu\tensor\utils.h?

Reply:
sure

Reply:
Can this be deleted now?

---
In reply to: [331652371](https://github.com/microsoft/onnxruntime/pull/1983#discussion_r331652371) [](ancestors = 331652371)